### ISSUE TITLE
grammar support for arrow functions is inconsistent

### ISSUE THREAD

**Issue by vith** [gandm/language-babel#27]
![screenshot from 2015-07-18 03 51 33](https://cloud.githubusercontent.com/assets/3265539/8761003/6e959372-2d00-11e5-90d5-e6a845292f1c.png)

The gray `=>` aren't getting any tags applied; the ones that are purple got `storage.type.function.arrow.js`. I agree with their coloring using `one-dark-syntax` because they're consistent with the `function` keyword, but shouldn't they technically be classified as operators?

**Comment by gandm** [comment]
On holiday at the moment but will look at this when I'm back. I appreciate all your PR's and feedback.

> but shouldn't they technically be classified as operators?

Depends. The ecmascript def is that it is a Punctuator and also a ArrowFunction. I prefer the later.

As indicated in my README, Most of the grammar (except JSX) is based upon [babel-sublime](https://github.com/babel/babel-sublime) which in turn was based upon some textmate grammar. It also has it's own theme built-in so there are going to be some scope names that don't always play well with Atom themes. Also, textmate grammar has significant problems which make many things difficult or maybe even impossible. I think the problem with grammar needs addressing. Either using a programmatic interface that allows me to use the AST from babel or using a better grammar such as the new sublime format. I have rasied an [issue](https://github.com/atom/first-mate/issues/50) to see if atom can suppport the sublime syntax. And others have raised the use of the first option.

**Comment by vith** [comment]
> The ecmascript def is that it is a Punctuator and also a ArrowFunction.

Fair enough. I tried to check some other grammars for languages that have similar lambda function syntax but a lot of what I found looked like broken grammars. `language-csharp`, `language-java`, and `language-swift` either failed to label their arrows at all or incorrectly labeled individual characters in the arrow as separate operators. It's possible the sample code I was trying needed to be wrapped in more boilerplate to be valid though.

On the other hand, `language-haskell` gave `keyword.other.big-arrow.haskell` and `keyword.other.arrow.haskell` for the arrows in
```haskell
arr :: (Arrow a) => (b -> c) -> a b c
```

But anyway, I'm mostly just concerned with getting these arrows labeled at all.

I checked out the new `.sublime-syntax` format you mentioned and I agree the scope stack feature would be really nice to have. When working with the textmate-style grammars I kept wishing I could reuse rules like in [pegjs](https://github.com/pegjs/pegjs) grammars, but I think pushing/popping scopes would accomplish the same thing.
